### PR TITLE
Feature/http client invoke method async

### PR DESCRIPTION
**Pull request by wimvdc**
# Description

This is not an ideal solution. I haven't found a working solution to return the "awaited" result when a loop is already running.
I have added a new optional parameter to `invoke_method` named `return_coroutine` which allows the method to return a coroutine.

Example usage:
```
with DaprClient() as dapr_client:
            result = await dapr_client.invoke_method(app_id="dummyapp",
                method_name="my_dummy",
                data=None,
                return_coroutine=True)
```

Please advice on this approach or suggest a better one.

Note: Might be related to #90 which tries to do something similar for the grpc side of the client.

## Issue reference

Please reference the issue this PR will close: #295 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

### CODE REVIEW

File: dapr/clients/http/dapr_invocation_http_client.py [hunk 1]
```diff
@@ -39,7 +39,8 @@ def invoke_method(
             content_type: Optional[str] = None,
             metadata: Optional[MetadataTuple] = None,
             http_verb: Optional[str] = None,
-            http_querystring: Optional[MetadataTuple] = None) -> InvokeMethodResponse:
```
Comment (wcs1only) [inline dapr/clients/http/dapr_invocation_http_client.py:42]: How about this? Can you just add an invoke_method_async call? I think having separate synchronous and async signatures might avoid confusion and maintain backwards compatibility.

Comment (wimvdc) [reply dapr/clients/http/dapr_invocation_http_client.py:42]: Thanks for the response. This does indeed sound like a better approach.
Can you review this updated PR?

**Comment by wimvdc** [comment]
Still waiting for some feedback

File: dapr/clients/__init__.py [hunk 2]
```diff
@@ -115,3 +115,47 @@ def invoke_method(
...
+                app_id,
+                method_name,
+                data,
+                content_type=content_type,
+                metadata=metadata,
+                http_verb=http_verb,
+                http_querystring=http_querystring,
+                return_coroutine=True)
+        else:
+            return super().invoke_method(
```
Comment (paulyuk) [inline dapr/clients/__init__.py:154]: Should this else case also be async by adding await?

e.g.
`return await super().invoke_method(

**Review by paulyuk — COMMENTED** [review]
Like the new design of asyncInvoke functions.  See comment with q on super() case.

File: dapr/clients/__init__.py (hunk 2 above)
Comment (paulyuk) [reply dapr/clients/__init__.py:154]:
```suggestion
            return await super().invoke_method(
```

**Comment by wcs1only** [comment]
My apologies, I tend to mentally filter out draft PRs when processing the backlog. I like your new approach. Do you consider this PR ready to be merged? If so, can you please mark it as "Ready for review" ?

File: dapr/clients/__init__.py (hunk 2 above)
Comment (wimvdc) [reply dapr/clients/__init__.py:162]: Firstly, thanks for the review and to answer your question/remark; no, I don't think so.
`super()` refers to `DaprGrpcClient`, the `invoke_method` there is completely synchronous. So as far my python knowledge goes, `await` it not needed here.

**Comment by wimvdc** [comment]
Fixed the linter issue which was reported (all linter issue should be resolved now)

**Comment by berndverst** [comment]
@wimvdc can you fix up the PR? The type check is failing

```
dapr/clients/__init__.py:151: error: Incompatible types in "await" (actual type "InvokeMethodResponse", expected type "Awaitable[Any]")  [misc]
dapr/clients/__init__.py:161: error: Incompatible return value type (got "InvokeMethodResponse", expected "Awaitable[InvokeMethodResponse]")  [return-value]
```

You can run this test with `tox -e mypy`

File: dapr/clients/__init__.py [hunk 3]
```diff
@@ -106,20 +106,70 @@ def invoke_method(
...
+            content_type (str, optional): Content type of the data.
+            metadata (MetadataTuple, optional): Additional metadata or headers.
+            http_verb (str, optional): HTTP verb for the request.
+            http_querystring (MetadataTuple, optional): Query parameters.
+
+        Returns:
+            Union[InvokeMethodResponse,Awaitable[Any]]: the response from the method invocation.
+        """
+        if self.invocation_client:
+            result = self.invocation_client.invoke_method(
```
Comment (jfilipe-trd) [inline dapr/clients/__init__.py:157]: Shouldn't be `result = await self.invocation_client.invoke_method` ? Otherwise the coroutine is never awaited.

**Comment by berndverst** [comment]
Now it's the linter
```bash
tox -e flake8
```

```
./dapr/clients/__init__.py:131:17: F541 f-string is missing placeholders
./dapr/clients/__init__.py:141:101: E501 line too long (109 > 100 characters)
./dapr/clients/http/dapr_invocation_http_client.py:51:74: E231 missing whitespace after ','
```

File: dapr/clients/__init__.py (hunk 3 above)
Comment (berndverst) [reply dapr/clients/__init__.py:157]:
```suggestion
            result = await self.invocation_client.invoke_method(
```

**Review by berndverst — CHANGES_REQUESTED** [review]
Please fix the linter warnings and take a look at the comment about awaiting the coroutine.

**Review by berndverst — DISMISSED** [review]

**Review by berndverst — CHANGES_REQUESTED** [review]
We cannot await a function that returns anything other than `Awaitable[Any]`

The mixed type approach doesn't work with our type hints. Can you find a way around that @wimvdc ?

**Comment by wimvdc** [comment]
> We cannot await a function that returns anything other than `Awaitable[Any]`
> 
> The mixed type approach doesn't work with our type hints. Can you find a way around that @wimvdc ?

I can try.
I'm currently involved with some other stuff so it might take a while until I have some free time to look at this.

**Comment by berndverst** [comment]
@wimvdc I rearchitected this PR quite a bit - but still keeping the essence of what you wanted. I really wanted to make sure your PR can make it into the the next release of the Python SDK!

**Review by berndverst — APPROVED** [review]

**Pull request merged**